### PR TITLE
Testing: Upgrade Jest to v27

### PR DESCRIPTION
**Pull request by gziolo**
## Description

Depends on #36328.

See details about breaking changes introduced in Jest v 27:
https://jestjs.io/blog/2021/05/25/jest-27

There is some bigger issue with unit tests that import `webpack` to build test fixtures. It either doesn't run at all or it takes forever and tests timeout.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/jest-preset-default/jest-preset.js
Comment (gziolo) [inline packages/jest-preset-default/jest-preset.js]: `false` is the default value so it can be safely removed. It only slows down test execution and it should never be enabled in the default preset.

Comment (gziolo) [reply packages/jest-preset-default/jest-preset.js]: Removed with https://github.com/WordPress/gutenberg/pull/33287/commits/cbf38a471eabe18f74fd7762b2088a4467fe9b2a

**Comment by gziolo** [comment]
The only remaining blocker is failing tests for React Native code. There is still something wrong with the configuration.

**Comment by simonhammes** [comment]
@gziolo I think I've managed to solve the issue of the JS unit tests timing out:

Apparently this has to do with a change in Jest concerning async hooks.

There are no timeouts after applying this patch (I know that you should generelly prefer async methods whenever IO is involved):

```patch
diff --git a/test/integration/full-content/full-content.test.js b/test/integration/full-content/full-content.test.js
index 64862d19f0..363169c857 100644
--- a/test/integration/full-content/full-content.test.js
+++ b/test/integration/full-content/full-content.test.js
@@ -58,8 +58,8 @@ function normalizeParsedBlocks( blocks ) {
 }
 
 describe( 'full post content fixture', () => {
-	beforeAll( async () => {
-		const blockMetadataFiles = await glob(
+	beforeAll( () => {
+		const blockMetadataFiles = glob.sync(
 			'packages/block-library/src/*/block.json'
 		);
 		const blockDefinitions = fromPairs(
```

**Comment by simonhammes** [comment]
And most (if not all of them) of the native test errors should be gone once RN is updated to 0.66 (https://github.com/WordPress/gutenberg/pull/36328), since that would include @jest/create-cache-key-function@27 (see https://github.com/callstack/react-native-testing-library/issues/743)

**Comment by gziolo** [comment]
> Apparently this has to do with a change in Jest concerning async hooks.
>
> There are no timeouts after applying this patch (I know that you should generelly prefer async methods whenever IO is involved)

Interesting discovery. Do you know what has changed in regards to async hooks? I would love to learn more.

**Comment by simonhammes** [comment]
> > Apparently this has to do with a change in Jest concerning async hooks.
> > There are no timeouts after applying this patch (I know that you should generelly prefer async methods whenever IO is involved)
> 
> Interesting discovery. Do you know what has changed in regards to async hooks? I would love to learn more.

No, sorry, I did not investigate this any further.

EDIT: I guess I could try and find a minimal example which reproduces this bug and open an issue over at facebook/jest.

**Comment by gziolo** [comment]
I rebased the branch with the most recent changes in `trunk`. There is now Jest `27.4.4` so I updated dependencies again. I also applied the fix that @simonhammes suggested in https://github.com/WordPress/gutenberg/pull/33287#issuecomment-983991349.

> And most (if not all of them) of the native test errors should be gone once RN is updated to 0.66 (#36328), since that would include @jest/create-cache-key-function@27 (see callstack/react-native-testing-library#743)

We are still waiting for React Native upgrade.

**Comment by geriux** [comment]
> We are still waiting for React Native upgrade.

FYI The [upgrade](https://github.com/WordPress/gutenberg/pull/36328) was just merged 🎉

**Comment by gziolo** [comment]
> > We are still waiting for React Native upgrade.
> 
> FYI The [upgrade](https://github.com/WordPress/gutenberg/pull/36328) was just merged 🎉

Awesome. I saw the commit earlier today. I will plan to bring this PR up to date this week. Thank you for making it happen 👏🏻

**Comment by geriux** [comment]
> > > We are still waiting for React Native upgrade.
> > 
> > 
> > FYI The [upgrade](https://github.com/WordPress/gutenberg/pull/36328) was just merged 🎉
> 
> Awesome. I saw the commit earlier today. I will plan to bring this PR up to date this week. Thank you for making it happen 👏🏻

Cool! Kudos to the people involved with the upgrade @dcalhoun @twstokes @mchowning @hypest 👏 let us know if you need any further help 👍

**Comment by gziolo** [comment]
I updated this branch with the latest changes in `trunk`. It looks like `npm run native test` is very slow now. It takes 2-3 minutes on CI for `trunk` (https://github.com/WordPress/gutenberg/runs/4699951957?check_suite_focus=true). While here it's over 20 minutes and it is still running. I would appreciate some help with debugging because it was working perfectly fine before the upgrade to RN 0.66 landed.

**Comment by dcalhoun** [comment]
@gziolo I am uncertain as to why the mobile tests are still failing after https://github.com/WordPress/gutenberg/pull/33287/commits/58725edcc6f2ffce4462ffe5fa9a5dd414b069da. They succeeded in #37715. 

However, I have seen the current `TypeError: Cannot use 'in' operator to search for 'WebkitAnimation' in undefined` error before. It previously originated from Enzyme forcing loading of `react-dom`, which fails to run after switching the test environment from `jsdom` to `node`. All usage of Enzyme was removed from `*.native.js` test files. I verified that to be the case in this branch for the currently failing files. 

I wonder if this may be a test environment caching issue? How might we ensure a fresh cache?

**Comment by gziolo** [comment]
I rebased the branch with the latest changes from `trunk` hoping it resets the cache, but it looks like I see the same failures locally:

Screenshot 2022-01-12 at 15 35 24

Is it possible that some changes on `trunk` got applied in the meantime which require JSDom to pass? WebkitAnimation is definitely not something you would expect in React Native.

**Comment by gziolo** [comment]
@dcalhoun, I updated a bunch of dev dependencies so maybe it's related to that. Anyway, it would be good to find out why the code would want to use `WebkitAnimation` in the first place. Maybe, it's something that needs an override for mobile apps.

**Comment by dcalhoun** [comment]
@gziolo I opened https://github.com/WordPress/gutenberg/pull/37921 to address the current native test failures.

**Comment by gziolo** [comment]
> @gziolo I opened #37921 to address the current native test failures.

It looks like it does the trick. I'm super excited about all the improvements coming with this PR to the native unit tests! I'd appreciate some sanity check for the web changes in this PR as well.

Separately from this PR we can look into:
- switching to `node` as the default test env for web unit tests instead of `jsdom` that I set for backward compatibility
- switching to the new test runner in native e2e tests

**Comment by gziolo** [comment]
@dcalhoun and @fluiddot – let's see if we finally get green light from CI 😄

**Comment by dcalhoun** [comment]
> switching to `node` as the default test env for web unit tests instead of `jsdom` that I set for backward compatibility

Is the target with this idea to avoid the significant performance overhead incurred by `jsdom`? Given the web tests are so DOM-centric, leveraging `jsdom` does make some sense. Would the alternative plan be to maintain mocks of the specific DOM APIs required? 

> switching to the new test runner in native e2e tests

Agreed. I considered doing this in https://github.com/WordPress/gutenberg/pull/37715, but opted to postpone that work to keep the PR focused purely on fixing the failing tests. Hopefully we can update the e2e tests soon as well.

**Comment by gziolo** [comment]
> > switching to node as the default test env for web unit tests instead of jsdom that I set for backward compatibility
>
> Is the target with this idea to avoid the significant performance overhead incurred by jsdom? Given the web tests are so DOM-centric, leveraging jsdom does make some sense. Would the alternative plan be to maintain mocks of the specific DOM APIs required?

It's not only performance but making it explicit where a test needs DOM. Overall, it would lead to a better design that works better with server-side rendering (https://github.com/WordPress/gutenberg/issues/17273) and React Native. You can always explicitly set [`testEnvironment`](https://jestjs.io/docs/configuration#testenvironment-string) to `jsdom` for individual tests with:

```js
/**
 * @jest-environment jsdom
 */
```

**Comment by gziolo** [comment]
I rebased this branch with `trunk` again. I see that there is a new test failing. It was introduced by @geriux in https://github.com/WordPress/gutenberg/pull/37951. @dcalhoun or @fluiddot, can you help to fix it? I'm able to reproduce it locally, but I can't figure out why it complains because `act` is used internally in `initializeEditor`:

Screenshot 2022-01-18 at 14 11 20

Once we have this test sorted out, I will merge this PR to avoid the hassle of bringing the branch up to date. I think it had enough testing from a few people already.

**Comment by fluiddot** [comment]
> I rebased this branch with `trunk` again. I see that there is a new test failing. It was introduced by @geriux in #37951. @dcalhoun or @fluiddot, can you help to fix it?

Thanks for raising this issue @gziolo 🙇 , I'll take a look at this as soon as possible 💨 .

**Comment by gziolo** [comment]
I hope to merge this PR as first thing in the morning. Let's hope nothing new doesn't pop up after I rebase with `trunk` 🤞🏻

File: packages/url/src/test/index.native.js
Comment (dcalhoun) [inline packages/url/src/test/index.native.js:49]: What is the purpose of these additions?

**Review by dcalhoun — APPROVED** [review]
🚀 LGTM. Thank you for the hard work to land this upgrade. Some really great improvements in this PR. 👏🏻

File: packages/url/src/test/index.native.js
Comment (gziolo) [reply packages/url/src/test/index.native.js:49]: The implementation of URL polyfill for React Native doesn’t fully align with the one used in the browser. The text fixture we downloaded passes all checks for web but we need to list exceptions for native implementation. If the list of exceptions grows too much we might need to revisit this approach.

**Comment by gziolo** [comment]
Great team work. Thank you all for all the help.

**Pull request merged**
